File: csharp/PistaDBTypes.cs
/**
 *    ___ _    _        ___  ___
 *   | _ (_)__| |_ __ _|   \| _ )
 *   |  _/ (_-<  _/ _` | |) | _ \
 *   |_| |_/__/\__\__,_|___/|___/
 *
 * PistaDB C# Binding — PistaDBTypes.cs
 * Public types: enums, result types, parameters, exception.
 */

using System;

namespace PistaDB
{
    // ── Distance metrics ──────────────────────────────────────────────────────

    /// <summary>Vector distance metric used by the database.</summary>
    public enum Metric
    {
        /// <summary>Euclidean distance (L2). Default; good for most embeddings.</summary>
        L2      = 0,
        /// <summary>Cosine similarity expressed as distance. Ideal for text embeddings.</summary>
        Cosine  = 1,
        /// <summary>Inner product (stored as negative). Good for dot-product similarity.</summary>
        IP      = 2,
        /// <summary>Manhattan distance (L1).</summary>
        L1      = 3,
        /// <summary>Hamming distance. Suitable for binary or integer vectors.</summary>
        Hamming = 4,
    }

    // ── Index algorithms ──────────────────────────────────────────────────────

    /// <summary>ANN index algorithm to use.</summary>
    public enum IndexType
    {
        /// <summary>Brute-force exact scan. Perfect accuracy, O(n) query.</summary>
        Linear  = 0,
        /// <summary>HNSW — fast approximate search. Recommended for RAG workloads.</summary>
        HNSW    = 1,
        /// <summary>Inverted File Index. Requires <see cref="PistaDatabase.Train"/> before insert.</summary>
        IVF     = 2,
        /// <summary>IVF + Product Quantization. Low memory, lossy. Requires Train.</summary>
        IVF_PQ  = 3,
        /// <summary>DiskANN / Vamana. Scales to billion-vector datasets.</summary>
        DiskANN = 4,
        /// <summary>Locality-Sensitive Hashing. Ultra-low memory footprint.</summary>
        LSH     = 5,
        /// <summary>ScaNN (Anisotropic Vector Quantization). Two-phase reranking.</summary>
        ScaNN   = 6,
    }

    // ── Result types ──────────────────────────────────────────────────────────

    /// <summary>A single KNN search result.</summary>
    public sealed class SearchResult
    {
        /// <summary>User-supplied vector id.</summary>
        public ulong Id { get; }
        /// <summary>Distance from the query (lower is more similar for L2/L1/Hamming/Cosine).</summary>
        public float Distance { get; }
        /// <summary>Optional human-readable label associated with the vector.</summary>
        public string Label { get; }

        internal SearchResult(ulong id, float distance, string label)
        {
            Id = id;
            Distance = distance;
            Label = label;
        }

        public override string ToString() =>
            $"SearchResult(Id={Id}, Distance={Distance:F4}, Label=\"{Label}\")";
    }

    /// <summary>A stored vector entry returned by <see cref="PistaDatabase.Get"/>.</summary>
    public sealed class VectorEntry
    {
        /// <summary>The raw float vector.</summary>
        public float[] Vector { get; }
        /// <summary>Optional label stored alongside the vector.</summary>
        public string Label { get; }

        internal VectorEntry(float[] vector, string label)
        {
            Vector = vector;
            Label  = label;
        }
    }

    // ── Parameters ────────────────────────────────────────────────────────────

    /// <summary>
    /// Index parameters passed when opening a database.
    /// Unused parameters for the selected <see cref="IndexType"/> are ignored.
    /// </summary>
    public sealed class PistaDBParams
    {
        // ── HNSW ──────────────────────────────────────────────────────────────

        /// <summary>Max bi-directional connections per layer. Higher = better recall, more memory. Default 16.</summary>
        public int   HnswM              { get; set; } = 16;
        /// <summary>Build-time search width. Higher = better graph quality, slower build. Default 200.</summary>
        public int   HnswEfConstruction { get; set; } = 200;
        /// <summary>Query-time search width. Higher = better recall, slower query. Default 50.</summary>
        public int   HnswEfSearch       { get; set; } = 50;

        // ── IVF / IVF_PQ ──────────────────────────────────────────────────────

        /// <summary>Number of IVF centroids (clusters). Default 128.</summary>
        public int   IvfNList  { get; set; } = 128;
        /// <summary>Number of centroids to search at query time. Default 8.</summary>
        public int   IvfNProbe { get; set; } = 8;
        /// <summary>Number of PQ sub-spaces. Default 8.</summary>
        public int   PqM       { get; set; } = 8;
        /// <summary>Bits per PQ sub-code (4 or 8). Default 8.</summary>
        public int   PqNBits   { get; set; } = 8;

        // ── DiskANN / Vamana ──────────────────────────────────────────────────

        /// <summary>Max graph out-degree. Default 32.</summary>
        public int   DiskannR     { get; set; } = 32;
        /// <summary>Build-time search list size. Default 100.</summary>
        public int   DiskannL     { get; set; } = 100;
        /// <summary>Pruning parameter (≥ 1.0). Default 1.2.</summary>
        public float DiskannAlpha { get; set; } = 1.2f;

        // ── LSH ───────────────────────────────────────────────────────────────

        /// <summary>Number of hash tables. Default 10.</summary>
        public int   LshL { get; set; } = 10;
        /// <summary>Hash functions per table. Default 8.</summary>
        public int   LshK { get; set; } = 8;
        /// <summary>Bucket width (E2LSH). Default 10.0.</summary>
        public float LshW { get; set; } = 10.0f;

        // ── ScaNN ─────────────────────────────────────────────────────────────

        /// <summary>Coarse IVF partitions. Default 128.</summary>
        public int   ScannNList   { get; set; } = 128;
        /// <summary>Partitions to probe during search. Default 32.</summary>
        public int   ScannNProbe  { get; set; } = 32;
        /// <summary>PQ sub-spaces. Default 8.</summary>
        public int   ScannPqM     { get; set; } = 8;
        /// <summary>Bits per PQ sub-code (4 or 8). Default 8.</summary>
        public int   ScannPqBits  { get; set; } = 8;
        /// <summary>Candidates to rerank with exact distances. Default 100.</summary>
        public int   ScannRerankK { get; set; } = 100;
        /// <summary>Anisotropic penalty η. Default 0.2.</summary>
        public float ScannAqEta   { get; set; } = 0.2f;

        // ── Presets ───────────────────────────────────────────────────────────

        /// <summary>High-recall preset: large M and ef values for HNSW.</summary>
        public static PistaDBParams HighRecall => new PistaDBParams
        {
            HnswM              = 32,
            HnswEfConstruction = 400,
            HnswEfSearch       = 200,
        };

        /// <summary>Low-latency preset: smaller ef values for HNSW.</summary>
        public static PistaDBParams LowLatency => new PistaDBParams
        {
            HnswM              = 16,
            HnswEfConstruction = 100,
            HnswEfSearch       = 20,
        };
    }

    // ── Exception ─────────────────────────────────────────────────────────────

    /// <summary>Thrown when a PistaDB native operation fails.</summary>
    public class PistaDBException : Exception
    {
        public PistaDBException(string message) : base(message) { }
        public PistaDBException(string message, Exception inner) : base(message, inner) { }
    }
}

File: wrap/ios/Sources/PistaDB/PistaDBTypes.swift
/*
 *    ___ _    _        ___  ___
 *   | _ (_)__| |_ __ _|   \| _ )
 *   |  _/ (_-<  _/ _` | |) | _ \
 *   |_| |_/__/\__\__,_|___/|___/
 */
import Foundation
import PistaDBObjC

// MARK: - Metric

/// Distance metric used for vector comparisons.
public enum Metric: Int, CaseIterable, Sendable {
    /// Euclidean distance – general purpose, image and multimodal embeddings.
    case l2      = 0
    /// Cosine similarity stored as `1 − similarity` – text embeddings.
    case cosine  = 1
    /// Inner product stored as negative dot product – normalised vectors.
    case ip      = 2
    /// Manhattan / L1 distance – sparse vectors.
    case l1      = 3
    /// Hamming distance – binary embeddings, deduplication.
    case hamming = 4

    var objc: PSTMetric { PSTMetric(rawValue: rawValue)! }

    init(_ objc: PSTMetric) { self.init(rawValue: objc.rawValue)! }
}

// MARK: - IndexType

/// Index algorithm used to organise and search vectors.
public enum IndexType: Int, CaseIterable, Sendable {
    /// Brute-force linear scan – exact results, no training needed.
    case linear  = 0
    /// Hierarchical Navigable Small World graphs (recommended default).
    case hnsw    = 1
    /// Inverted File Index – requires ``PistaDB/train()`` before inserting.
    case ivf     = 2
    /// IVF + Product Quantisation – requires ``PistaDB/train()`` before inserting.
    case ivfPQ   = 3
    /// Vamana / DiskANN – billion-scale datasets.
    case diskANN = 4
    /// Locality-Sensitive Hashing – ultra-low memory.
    case lsh     = 5
    /// ScaNN: Anisotropic Vector Quantisation – requires ``PistaDB/train()``.
    case scaNN   = 6

    var objc: PSTIndexType { PSTIndexType(rawValue: rawValue)! }

    init(_ objc: PSTIndexType) { self.init(rawValue: objc.rawValue)! }
}

// MARK: - SearchResult

/// A single k-NN result returned by ``PistaDB/search(_:k:)``.
/// Results are ordered by ascending distance (nearest first).
public struct SearchResult: Sendable {
    /// The vector's unique identifier.
    public let id: UInt64
    /// Distance from the query (interpretation depends on metric).
    public let distance: Float
    /// Human-readable label supplied at insert time. Empty if none.
    public let label: String

    init(_ pst: PSTSearchResult) {
        self.id       = pst.vectorId
        self.distance = pst.distance
        self.label    = pst.label
    }
}

// MARK: - VectorEntry

/// A stored vector with its label, returned by ``PistaDB/get(id:)``.
public struct VectorEntry: Sendable {
    /// The raw float vector (length == `dim`).
    public let vector: [Float]
    /// Human-readable label supplied at insert time. Empty if none.
    public let label: String

    init(_ pst: PSTVectorEntry) {
        self.label  = pst.label
        self.vector = pst.vectorData.withUnsafeBytes { raw in
            Array(raw.bindMemory(to: Float.self))
        }
    }
}

// MARK: - PistaDBParams

/// Index tuning parameters passed when opening a database.
///
/// ```swift
/// var p = PistaDBParams()
/// p.hnswM = 32
/// p.hnswEfSearch = 100
/// ```
public struct PistaDBParams: Sendable {

    // HNSW
    public var hnswM:              Int   = 16
    public var hnswEfConstruction: Int   = 200
    public var hnswEfSearch:       Int   = 50

    // IVF / IVF_PQ
    public var ivfNlist:           Int   = 128
    public var ivfNprobe:          Int   = 8
    public var pqM:                Int   = 8
    public var pqNbits:            Int   = 8

    // DiskANN
    public var diskannR:           Int   = 32
    public var diskannL:           Int   = 100
    public var diskannAlpha:       Float = 1.2

    // LSH
    public var lshL:               Int   = 10
    public var lshK:               Int   = 8
    public var lshW:               Float = 10.0

    // ScaNN
    public var scannNlist:         Int   = 128
    public var scannNprobe:        Int   = 32
    public var scannPqM:           Int   = 8
    public var scannPqBits:        Int   = 8
    public var scannRerankK:       Int   = 100
    public var scannAqEta:         Float = 0.2

    public init() {}

    func toPST() -> PSTParams {
        let p = PSTParams.default()
        p.hnswM               = hnswM
        p.hnswEfConstruction  = hnswEfConstruction
        p.hnswEfSearch        = hnswEfSearch
        p.ivfNlist            = ivfNlist
        p.ivfNprobe           = ivfNprobe
        p.pqM                 = pqM
        p.pqNbits             = pqNbits
        p.diskannR            = diskannR
        p.diskannL            = diskannL
        p.diskannAlpha        = diskannAlpha
        p.lshL                = lshL
        p.lshK                = lshK
        p.lshW                = lshW
        p.scannNlist          = scannNlist
        p.scannNprobe         = scannNprobe
        p.scannPqM            = scannPqM
        p.scannPqBits         = scannPqBits
        p.scannRerankK        = scannRerankK
        p.scannAqEta          = scannAqEta
        return p
    }
}

// MARK: - PistaDBError

/// Errors thrown by ``PistaDB`` operations.
public enum PistaDBError: LocalizedError, Sendable {
    case openFailed(String)
    case notFound
    case alreadyExists
    case invalidArgument(String)
    case notTrained
    case ioError(String)
    case corrupt
    case closed
    case operationFailed(String)

    public var errorDescription: String? {
        switch self {
        case .openFailed(let m):     return "Failed to open database: \(m)"
        case .notFound:              return "Vector not found"
        case .alreadyExists:         return "Vector ID already exists"
        case .invalidArgument(let m): return "Invalid argument: \(m)"
        case .notTrained:            return "Index not trained – call train() before inserting"
        case .ioError(let m):        return "I/O error: \(m)"
        case .corrupt:               return "Database file is corrupt"
        case .closed:                return "Database is already closed"
        case .operationFailed(let m): return "Operation failed: \(m)"
        }
    }
}

// MARK: - Internal helpers

extension NSError {
    func toPistaDBError() -> PistaDBError {
        guard domain == PSTDatabaseErrorDomain else {
            return .operationFailed(localizedDescription)
        }
        // Swift strips the common "PSTDatabaseError" prefix from NS_ENUM cases:
        // PSTDatabaseErrorNotFound → PSTDatabaseErrorCode.notFound
        switch PSTDatabaseErrorCode(rawValue: code) {
        case .notFound:      return .notFound
        case .alreadyExists: return .alreadyExists
        case .notTrained:    return .notTrained
        case .corrupt:       return .corrupt
        case .io:            return .ioError(localizedDescription)
        case .invalidArg:    return .invalidArgument(localizedDescription)
        case .noMemory:      return .operationFailed("Out of memory")
        default:             return .operationFailed(localizedDescription)
        }
    }
}
